core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryPresenter.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.core.history
+
+import android.util.Log
+import io.reactivex.CompletableObserver
+import io.reactivex.Observable
+import io.reactivex.Scheduler
+import io.reactivex.SingleObserver
+import io.reactivex.disposables.Disposable
+import org.kiwix.kiwixmobile.core.base.BasePresenter
+import org.kiwix.kiwixmobile.core.data.DataSource
+import org.kiwix.kiwixmobile.core.di.qualifiers.Computation
+import org.kiwix.kiwixmobile.core.di.qualifiers.MainThread
+import org.kiwix.kiwixmobile.core.history.HistoryContract.Presenter
+import org.kiwix.kiwixmobile.core.history.HistoryContract.View
+import org.kiwix.kiwixmobile.core.history.HistoryListItem.HistoryItem
+import javax.inject.Inject
+
+internal class HistoryPresenter @Inject constructor(
+  private val dataSource: DataSource,
+  @param:MainThread private val mainThread: Scheduler,
+  @param:Computation private val computation: Scheduler
+) : BasePresenter<View>(), Presenter {
+  private var disposable: Disposable? = null
+
+  override fun loadHistory(showHistoryCurrentBook: Boolean) {
+    dataSource.getDateCategorizedHistory(showHistoryCurrentBook)
+      .subscribe(object : SingleObserver<List<HistoryListItem>> {
+        override fun onSubscribe(d: Disposable) {
+          if (disposable != null && !disposable!!.isDisposed) {
```

Comment:
May be you can use this then
` if (!disposable?.isDisposed!!) {
            disposable!!.dispose()
          }'
@macgills  can better suggest

Reply:
Don't use `!!`.
Try going through the merged PRs for converting code to Kotlin you'll find some useful comments there which can be applied here as well. This will save your time.

Reply:
It is quite annoying code, I think the intention here is if there is an already running load operation then cancel and start a new one when we receive a fresh request. Our SearchViewModel does precisely this with a handy little operator called switchMap

```
  private fun searchResultsFromZimReader() = filter
    .distinctUntilChanged()
    .switchMap(::searchResults)
```